File: backend/app/api/v1/jobs.py
from fastapi import APIRouter, Depends, Query, status, HTTPException
from sqlalchemy.ext.asyncio import AsyncSession
import uuid
import logging
from typing import List, Dict, Any
import json
from pydantic import BaseModel, Field
import httpx

from app.core.database import get_db
from app.core.dependencies import get_current_user, get_job_discovery_service, get_ai_provider
from app.models.user import User
from app.schemas.job import (
    JobSearchResponse, JobSearchResponseData,
    JobDetailResponse, JobDetailData,
    BookmarkActionResponse, BookmarkActionData,
    BookmarkListResponse, BookmarkListData,
    SavedSearchRequest, SavedSearchResponse, SavedSearchData,
    SavedSearchListResponse, SavedSearchListData,
    RecommendationListResponse, RecommendationListData,
)
from app.services.interfaces.job_discovery_service import IJobDiscoveryService
from app.core.config import settings

logger = logging.getLogger(__name__)


router = APIRouter(
    prefix="/jobs",
    tags=["Job Discovery - Khám phá việc làm"]
)

@router.get(
    "/search",
    response_model=JobSearchResponse,
    summary="Tìm kiếm việc làm",
    description="Tìm kiếm việc làm qua JSearch API (có cache 4 giờ), hỗ trợ lọc theo địa điểm, mức lương, hình thức làm việc và phân trang."
)
async def search_jobs(
    q: str = Query(..., description="Từ khóa tìm kiếm (chức danh, kỹ năng)"),
    location: str | None = Query(default=None),
    employment_type: str | None = Query(default=None, description="remote, hybrid, on-site"),
    experience_level: str | None = Query(default=None, description="entry, mid, senior"),
    salary_min: int | None = Query(default=None, ge=0),
    remote: bool = Query(default=False),
    page: int = Query(default=1, ge=1),
    per_page: int = Query(default=20, ge=1, le=50),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    service: IJobDiscoveryService = Depends(get_job_discovery_service),
) -> JobSearchResponse:
    result = await service.search_jobs(
        user_id=current_user.id,
        query=q,
        location=location,
        employment_type=employment_type,
        experience_level=experience_level,
        salary_min=salary_min,
        remote_only=remote,
        page=page,
        per_page=per_page,
    )
    await db.commit()
    return JobSearchResponse(success=True, data=JobSearchResponseData(**result))


@router.get(
    "/bookmarks",
    response_model=BookmarkListResponse,
    summary="Danh sách việc làm đã lưu",
    description="Lấy danh sách các tin tuyển dụng người dùng đã bookmark."
)
async def get_bookmarks(
    current_user: User = Depends(get_current_user),
    service: IJobDiscoveryService = Depends(get_job_discovery_service),
) -> BookmarkListResponse:
    bookmarks = await service.get_bookmarks(user_id=current_user.id)
    return BookmarkListResponse(
        success=True,
        data=BookmarkListData(bookmarks=bookmarks, total=len(bookmarks)),
    )


@router.get(
    "/recommendations",
    response_model=RecommendationListResponse,
    summary="Gợi ý việc làm theo hồ sơ",
    description="Gợi ý các việc làm phù hợp nhất dựa trên kỹ năng trong hồ sơ ứng viên."
)
async def get_recommendations(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    service: IJobDiscoveryService = Depends(get_job_discovery_service),
) -> RecommendationListResponse:
    recs = await service.get_recommendations(user_id=current_user.id)
    await db.commit()
    return RecommendationListResponse(
        success=True,
        data=RecommendationListData(recommendations=recs, total=len(recs)),
    )


@router.get(
    "/saved-searches",
    response_model=SavedSearchListResponse,
    summary="Danh sách tìm kiếm đã lưu",
)
async def list_saved_searches(
    current_user: User = Depends(get_current_user),
    service: IJobDiscoveryService = Depends(get_job_discovery_service),
) -> SavedSearchListResponse:
    searches = await service.list_saved_searches(user_id=current_user.id)
    return SavedSearchListResponse(
        success=True,
        data=SavedSearchListData(saved_searches=searches, total=len(searches)),
    )


@router.post(
    "/saved-searches",
    response_model=SavedSearchResponse,
    status_code=status.HTTP_201_CREATED,
    summary="Lưu bộ tiêu chí tìm kiếm",
    description="Lưu lại tiêu chí tìm kiếm để nhận thông báo khi có việc làm mới phù hợp."
)
async def create_saved_search(
    payload: SavedSearchRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    service: IJobDiscoveryService = Depends(get_job_discovery_service),
) -> SavedSearchResponse:
    saved = await service.create_saved_search(
        user_id=current_user.id,
        name=payload.name,
        criteria=payload.search_criteria,
        alert_frequency=payload.alert_frequency,
    )
    await db.commit()
    return SavedSearchResponse(success=True, data=SavedSearchData(**saved))


@router.get(
    "/{job_id}",
    response_model=JobDetailResponse,
    summary="Chi tiết việc làm",
    description="Xem thông tin chi tiết một tin tuyển dụng và ghi nhận lượt xem."
)
async def get_job_details(
    job_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    service: IJobDiscoveryService = Depends(get_job_discovery_service),
) -> JobDetailResponse:
    job = await service.get_job_details(user_id=current_user.id, job_id=uuid.UUID(job_id))
    await db.commit()
    return JobDetailResponse(success=True, data=JobDetailData(**job))


@router.post(
    "/{job_id}/bookmark",
    response_model=BookmarkActionResponse,
    summary="Lưu việc làm",
)
async def bookmark_job(
    job_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    service: IJobDiscoveryService = Depends(get_job_discovery_service),
) -> BookmarkActionResponse:
    await service.bookmark_job(user_id=current_user.id, job_id=uuid.UUID(job_id))
    await db.commit()
    return BookmarkActionResponse(
        success=True,
        data=BookmarkActionData(message="Đã lưu việc làm", is_bookmarked=True),
    )


@router.delete(
    "/{job_id}/bookmark",
    response_model=BookmarkActionResponse,
    summary="Bỏ lưu việc làm",
)
async def unbookmark_job(
    job_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
    service: IJobDiscoveryService = Depends(get_job_discovery_service),
) -> BookmarkActionResponse:
    await service.unbookmark_job(user_id=current_user.id, job_id=uuid.UUID(job_id))
    await db.commit()
    return BookmarkActionResponse(
        success=True,
        data=BookmarkActionData(message="Đã bỏ lưu việc làm", is_bookmarked=False),
    )


# --- Preserved: AI extract-requirements endpoint from feat-matching (kept) ---
class ExtractRequest(BaseModel):
    job_description: str = Field(..., description="Raw text of the job description.")

@router.post(
    "/extract-requirements",
    response_model=List[Dict[str, Any]],
    summary="Trích xuất yêu cầu công việc từ mô tả thô (JD)",
    description="Sử dụng AI để phân tích mô tả công việc thô và trích xuất danh sách các yêu cầu kỹ thuật có cấu trúc."
)
async def extract_requirements(
    payload: ExtractRequest,
    current_user: User = Depends(get_current_user)
) -> List[Dict[str, Any]]:
    api_key = settings.GEMINI_API_KEY
    if not api_key:
        # Fallback mocks
        return [
            {
                "id": "REQ1",
                "text": "Experience with Backend Development",
                "canonical": "Backend Development",
                "category": "Backend Development",
                "section": "must_have",
                "priority": "CRITICAL"
            },
            {
                "id": "REQ2",
                "text": "Knowledge of SQL Database",
                "canonical": "SQL Database",
                "category": "SQL Database",
                "section": "required",
                "priority": "HIGH"
            }
        ]

    prompt = (
        "You are an expert technical recruiter. Analyze the following job description text "
        "and extract a list of structured requirements. "
        "Return strictly a JSON array of objects. Do not include markdown formatting (like ```json) or any explanations.\n\n"
        "Each object in the array must strictly have these fields:\n"
        "- 'id': unique string identifier (e.g. 'REQ1', 'REQ2')\n"
        "- 'text': raw requirement sentence from JD\n"
        "- 'canonical': canonical technical concept (e.g. 'Backend Development', 'Containerization', 'SQL Database', 'Frontend Development')\n"
        "- 'category': high-level category\n"
        "- 'section': must be one of ['must_have', 'required', 'preferred', 'nice_to_have']\n"
        "- 'priority': must be one of ['CRITICAL', 'HIGH', 'MEDIUM', 'LOW']\n\n"
        f"Job Description:\n{payload.job_description}"
    )

    url = f"https://generativelanguage.googleapis.com/v1beta/models/gemini-2.5-flash:generateContent?key={api_key}"
    headers = {"Content-Type": "application/json"}
    body = {
        "contents": [{"parts": [{"text": prompt}]}],
        "generationConfig": {
            "responseMimeType": "application/json"
        }
    }

    async with httpx.AsyncClient() as client:
        try:
            response = await client.post(url, json=body, headers=headers, timeout=15.0)
            response.raise_for_status()
            res_data = response.json()
            content = res_data["candidates"][0]["content"]["parts"][0]["text"]
            return json.loads(content)
        except Exception as e:
            logger.error(f"Failed to extract requirements via Gemini: {e}")
            # Fallback mocks on error
            return [
                {
                    "id": "REQ1",
                    "text": "Experience with software engineering principles",
                    "canonical": "Backend Development",
                    "category": "Software",
                    "section": "must_have",
                    "priority": "CRITICAL"
                }
            ]

File: frontend/src/lib/services/jobs.ts
import api from "@/lib/axios";

export interface JobSummary {
  id: string;
  externalId: string;
  title: string;
  company: string;
  companyLogoUrl: string | null;
  location: string | null;
  employmentType: string | null;
  experienceLevel: string | null;
  salaryRange: string | null;
  salaryMin: number | null;
  salaryMax: number | null;
  skillsRequired: string[];
  postedDate: string | null;
  applicationUrl: string | null;
  isBookmarked: boolean;
}

export interface JobDetail extends JobSummary {
  description: string | null;
  requirements: string | null;
  benefits: string | null;
  industry: string | null;
  sourcePlatform: string | null;
  jobType: string | null;
}

export interface Pagination {
  page: number;
  perPage: number;
  total: number;
  totalPages: number;
}

export interface JobSearchResult {
  jobs: JobSummary[];
  pagination: Pagination;
}

export interface JobSearchFilters {
  q: string;
  location?: string;
  employmentType?: string;
  experienceLevel?: string;
  salaryMin?: number;
  remote?: boolean;
  page?: number;
  perPage?: number;
}

export interface BookmarkItem {
  job: JobSummary;
  bookmarkedAt: string;
}

export interface JobRecommendation {
  job: JobSummary;
  recommendationScore: number;
  recommendationReason: string;
}

export interface SavedSearch {
  id: string;
  name: string;
  searchCriteria: Record<string, unknown>;
  alertFrequency: string;
  isActive: boolean;
  createdAt: string;
}

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const mapJob = (j: any): JobDetail => ({
  id: j.id,
  externalId: j.external_id,
  title: j.title,
  company: j.company,
  companyLogoUrl: j.company_logo_url ?? null,
  location: j.location ?? null,
  employmentType: j.employment_type ?? null,
  experienceLevel: j.experience_level ?? null,
  salaryRange: j.salary_range ?? null,
  salaryMin: j.salary_min ?? null,
  salaryMax: j.salary_max ?? null,
  skillsRequired: (j.skills_required || []).map((s: unknown) => String(s)),
  postedDate: j.posted_date ?? null,
  applicationUrl: j.application_url ?? null,
  isBookmarked: !!j.is_bookmarked,
  description: j.description ?? null,
  requirements: j.requirements ?? null,
  benefits: j.benefits ?? null,
  industry: j.industry ?? null,
  sourcePlatform: j.source_platform ?? null,
  jobType: j.job_type ?? null,
});

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const mapPagination = (p: any): Pagination => ({
  page: p.page,
  perPage: p.per_page,
  total: p.total,
  totalPages: p.total_pages,
});

export const jobService = {
  async search(filters: JobSearchFilters): Promise<JobSearchResult> {
    const params: Record<string, string | number | boolean> = { q: filters.q };
    if (filters.location) params.location = filters.location;
    if (filters.employmentType) params.employment_type = filters.employmentType;
    if (filters.experienceLevel) params.experience_level = filters.experienceLevel;
    if (filters.salaryMin != null) params.salary_min = filters.salaryMin;
    if (filters.remote) params.remote = true;
    params.page = filters.page ?? 1;
    params.per_page = filters.perPage ?? 20;

    const response = await api.get("/api/v1/jobs/search", { params });
    const d = response.data.data;
    return { jobs: (d.jobs || []).map(mapJob), pagination: mapPagination(d.pagination) };
  },

  async getJob(jobId: string): Promise<JobDetail> {
    const response = await api.get(`/api/v1/jobs/${jobId}`);
    return mapJob(response.data.data);
  },

  async bookmark(jobId: string): Promise<void> {
    await api.post(`/api/v1/jobs/${jobId}/bookmark`);
  },

  async unbookmark(jobId: string): Promise<void> {
    await api.delete(`/api/v1/jobs/${jobId}/bookmark`);
  },

  async getBookmarks(): Promise<BookmarkItem[]> {
    const response = await api.get("/api/v1/jobs/bookmarks");
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return (response.data.data.bookmarks || []).map((b: any) => ({
      job: mapJob(b.job),
      bookmarkedAt: b.bookmarked_at,
    }));
  },

  async getRecommendations(): Promise<JobRecommendation[]> {
    const response = await api.get("/api/v1/jobs/recommendations");
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return (response.data.data.recommendations || []).map((r: any) => ({
      job: mapJob(r.job),
      recommendationScore:
        r.recommendation_score != null ? parseFloat(r.recommendation_score) : 0,
      recommendationReason: r.recommendation_reason,
    }));
  },

  async listSavedSearches(): Promise<SavedSearch[]> {
    const response = await api.get("/api/v1/jobs/saved-searches");
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    return (response.data.data.saved_searches || []).map((s: any) => ({
      id: s.id,
      name: s.name,
      searchCriteria: s.search_criteria,
      alertFrequency: s.alert_frequency,
      isActive: s.is_active,
      createdAt: s.created_at,
    }));
  },

  async createSavedSearch(
    name: string,
    criteria: Record<string, unknown>,
    alertFrequency = "daily"
  ): Promise<SavedSearch> {
    const response = await api.post("/api/v1/jobs/saved-searches", {
      name,
      search_criteria: criteria,
      alert_frequency: alertFrequency,
    });
    const s = response.data.data;
    return {
      id: s.id,
      name: s.name,
      searchCriteria: s.search_criteria,
      alertFrequency: s.alert_frequency,
      isActive: s.is_active,
      createdAt: s.created_at,
    };
  },
};
